File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <bits/stdc++.h>
using namespace std;

#define ios cin.tie(NULL), ios_base::sync_with_stdio(false)
#define endl '\n'

int main(){
  ios; 

  while(1) {
    string s;
    string ans = "";
    getline(cin, s);

    if(s.size() == 1 && s[0]=='.') {
      break;
    }

    stack<char> stk;
    int size = s.size();
    bool found = false;
    for(int i=0; i<size; ++i) {
      if((s[i] == ')' || s[i] == ']') && stk.empty()) {
        cout << "no" << endl;
        found = true;
        break;
      }

      if(s[i]=='(' || s[i]=='[') {
        stk.push(s[i]);
      } else if(s[i]==')') {
        if(stk.empty() || stk.top() == '[') {
          cout << "no" << endl;
        found = true;
          break;
        } else {
          stk.pop();
        }
      } else if (s[i] == ']') {
        if(stk.empty() || stk.top() == '(') {
          cout << "no" << endl;
        found = true;
          break;
        } else {
          stk.pop();
        }
      }
    }

    if(!found) {
      cout << ( (stk.empty()) ? "yes" : "no" ) << endl;
    }
  }

  return 0;
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.rb
def balance?(str) 
	stack = []
	
	str.each_char { |c|
		if c=='(' or c=='[' then stack.push(c) end 
		if c==')'
			if(stack.empty? or stack[-1] == '[')
				return "no"
			else
				stack.pop
			end
		elsif c==']'
			if(stack.empty? or stack[-1] == '(')
				return "no"
			else 
				stack.pop
			end
		end
	}

	if stack.empty? then return "yes" else "no" end
end

loop {
	str = gets.chomp 
	if str=='.' then break end
	puts balance?(str)
}